// ng2-oauth2/src/provider/oauth2.provider.ts
import { Oauth2Service } from '../service/oauth2.service';
import { Oauth2Guard } from '../guard/oauth2.guard';
import { Oauth2AuthHttp } from '../service/oauth2.auth-http';
import { Router }    from '@angular/router';
import { provide } from '@angular/core';
import { Http } from '@angular/http';

export * from '../service/oauth2.service';
export * from '../service/oauth2.auth-http';
export * from '../service/lib/oauth2.id-token';
export * from '../service/lib/oauth2.access-token';
export * from '../service/lib/oauth2.oidc-config';
export * from '../service/lib/oauth2.profile';
export * from '../service/lib/oauth2.storage';

export const OAUTH2_PROVIDERS: any[] = [
    provide(
        Oauth2Service, {
            useClass: Oauth2Service,
            deps: [Http, Location],
            useFactory: oauth2ServiceFactory
        }
    ),
    provide(
        Oauth2Guard, {
            useClass: Oauth2Guard,
            deps: [Oauth2Service, Router]
        }
    ),
    provide(
        Oauth2AuthHttp, {
            useClass: Oauth2AuthHttp,
            deps: [Oauth2Service, Http]
        }
    )
];

export function oauth2ServiceFactory(location: Location, http: Http): Oauth2Service {
  return new Oauth2Service(location, http);
}

// for angular-cli
export default {
    providers: [Oauth2Service, Oauth2Guard, Oauth2AuthHttp]
};
